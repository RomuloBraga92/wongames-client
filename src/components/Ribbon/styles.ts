import styled, { css, DefaultTheme } from 'styled-components'
import { RibbonProps, RibbonColors } from '.'
import { darken } from 'polished'

export type RibbonContainerProps = Pick<RibbonProps, 'color' | 'size'>

const containerModifiers = {
  color: (theme: DefaultTheme, color: RibbonColors) => css`
    background-color: ${theme.colors[color]};

    &::before {
      border-left-color: ${darken(0.2, theme.colors[color])};
      border-top-color: ${darken(0.2, theme.colors[color])};
    }
  `,

  normal: (theme: DefaultTheme) => css`
    height: 3.6rem;
    font-size: ${theme.font.sizes.small};
    padding: 0 ${theme.spacings.small};
    right: -2rem;

    &::before {
      top: 3.6rem;
      border-right-width: 2rem;
      border-top-width: 1rem;
    }
  `,

  small: (theme: DefaultTheme) => css`
    height: 2.4rem;
    font-size: ${theme.font.sizes.xsmall};
    padding: 0 ${theme.spacings.xsmall};
    right: -1.6rem;

    &::before {
      top: 2.4rem;
      border-right-width: 1.6rem;
      border-top-width: 0.7rem;
    }
  `,

  large: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.large};
    padding: 0 ${theme.spacings.medium};
    right: -2.4rem;

    &::before {
      top: 4rem;
      border-right-width: 2.4rem;
      border-top-width: 1.2rem;
    }
  `,
}

export const Container = styled.div<RibbonContainerProps>`
  ${({ theme, color, size }) => css`
    display: flex;
    align-items: center;
    position: absolute;
    right: 0;
    color: ${theme.colors.white};
    font-weight: ${theme.font.bold};
    top: ${theme.spacings.small};

    &::before {
      content: '';
      position: absolute;
      right: 0;
      border-style: solid;
      border-left-width: 0rem;
      border-right-color: transparent;
      border-bottom-color: transparent;
      border-bottom-width: 1rem;
    }

    ${!!size && containerModifiers[size](theme)}
    ${!!color && containerModifiers.color(theme, color)}
  `}
`
