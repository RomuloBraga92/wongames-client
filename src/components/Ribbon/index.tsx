import * as S from './styles'

const Ribbon = () => (
  <S.Container>
    <h1>Ribbon</h1>
  </S.Container>
)

export default Ribbon
