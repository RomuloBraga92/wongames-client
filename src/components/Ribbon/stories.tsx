import { Story, Meta } from '@storybook/react/types-6-0'
import Ribbon from './'

export default {
  title: 'Ribbon',
  component: Ribbon,
} as Meta

export const Default: Story = () => <Ribbon />
