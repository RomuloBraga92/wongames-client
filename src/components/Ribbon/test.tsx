import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import Ribbon from './'

describe('<Ribbon />', () => {
  it('should render the text correctly', () => {
    renderWithTheme(<Ribbon>Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toBeInTheDocument()
  })

  it('should render a primary color by default', () => {
    renderWithTheme(<Ribbon>Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toHaveStyle({
      'background-color': '#F231A5',
    })
  })

  it('should render a secondary color when passed', () => {
    renderWithTheme(<Ribbon color="secondary">Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toHaveStyle({
      'background-color': '#3CD3C1',
    })
  })

  it('should render a normal size by default', () => {
    renderWithTheme(<Ribbon>Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toHaveStyle({
      'font-size': '1.4rem',
      height: '3.6rem',
    })
  })

  it('should render a small size when passed', () => {
    renderWithTheme(<Ribbon size="small">Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toHaveStyle({
      'font-size': '1.2rem',
      height: '2.4rem',
    })
  })

  it('should render a large size when passed', () => {
    renderWithTheme(<Ribbon size="large">Best Seller</Ribbon>)

    expect(screen.getByText(/best seller/i)).toHaveStyle({
      'font-size': '1.8rem',
      height: '4rem',
    })
  })
})
