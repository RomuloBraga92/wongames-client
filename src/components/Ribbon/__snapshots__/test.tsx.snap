// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Ribbon /> should render the heading 1`] = `
<main
  class=""
>
  <h1>
    Ribbon
  </h1>
</main>
`
